core/channel.go

```diff
@@ -416,3 +434,36 @@ func (ch *Channel) Reset() {
 		}
 	}
 }
+
+type idPool struct {
+	ids  []uint16
+	lock sync.Mutex
+}
+
+func newIDPool(maxID uint16) *idPool {
+	ids := make([]uint16, maxID)
+	for i := uint16(0); i < maxID; i++ {
+		ids[i] = i + 1
+	}
+	return &idPool{ids: ids}
+}
+
+func (p *idPool) Get() (uint16, error) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if len(p.ids) == 0 {
+		return 0, errors.New("no more IDs available")
+	}
+
+	id := p.ids[0]
+	p.ids = p.ids[1:]
+
+	return id, nil
+}
+
+func (p *idPool) Put(id uint16) {
+	p.lock.Lock()
+	p.ids = append(p.ids, id)
+	p.lock.Unlock()
+}
```

Comment:
I would rather go with a `freelist` and single `next` value in order to be keener on memory, e.g. below [0] the snippet I'm using in another project.

Also this logic could go directly in the `genericpool.New[*Channel](func() *Channel { ... }` function and leverage the `Reset()` callback for freeing the IDs.

[0]
````golang
type IdPool struct {
	lock     sync.Mutex
	nextFree uint32
	freeList []uint32
}

func NewIdPoolFrom(v uint32) IdPool { return IdPool{nextFree: v} }

func (self *IdPool) Allocate() (v uint32) {
	self.lock.Lock()
	defer self.lock.Unlock()
	if len(self.freeList) > 0 {
		v = self.freeList[len(self.freeList)-1]
		self.freeList = self.freeList[:len(self.freeList)-1]
	} else {
		v = self.nextFree
		self.nextFree++
	}
	return
}

func (self *IdPool) Free(v uint32) {
	self.lock.Lock()
	defer self.lock.Unlock()
	if v == self.nextFree {
		self.nextFree--
	} else {
		self.freeList = append(self.freeList, v)
	}
}
````

Reply:
That's neat. I will update the implementation.

Reply:
I have some questions about your implementation. Is there a reason why it does not handle maximum ID? The `Allocate` should return an error if all IDs are being used.

Reply:
I think it's just it was not done 😄 
iirc this was implemented quite quickly with the strong belief that `^uint32(0)` is infinity 😆

Reply:
I ended up with this implementation:

```go

type IdPool struct {
	lock     sync.Mutex
	nextFree uint16
	freeList []uint16
	maxID    uint16
}

func NewIDPool(max uint16) *IdPool {
	return &IdPool{
		freeList: make([]uint16, 0, max),
		maxID:    max,
	}
}

func (pool *IdPool) Get() (v uint16, err error) {
	pool.lock.Lock()
	defer pool.lock.Unlock()

	if len(pool.freeList) > 0 {
		v = pool.freeList[len(pool.freeList)-1]
		pool.freeList = pool.freeList[:len(pool.freeList)-1]
	} else if pool.nextFree > pool.maxID {
		return 0, errors.New("no more IDs available")
	} else {
		v = pool.nextFree
		pool.nextFree++
	}
	return
}

func (pool *IdPool) Put(v uint16) {
	pool.lock.Lock()
	defer pool.lock.Unlock()

	if v == pool.nextFree {
		pool.nextFree--
	} else {
		pool.freeList = append(pool.freeList, v)
	}
}
```

However I am not sure if the pre-allocation of the slice with IDs does not go against the "keen on memory" property you mentioned. I tried to run perf-bench example on both implementation an the results are not conclusive. Is the pre-allocation of slice for available IDs the memory you believe should be allocated over time?

Reply:
Thanks for doing the testing!
Ok, agreed if this does not show a clear gain on memory, we will probably be better off with the fix in its current state. Let's plan on refactoring this later !